src/core/lib/security/transport/security_connector.c

```diff
@@ -868,42 +966,55 @@ grpc_security_status grpc_ssl_channel_security_connector_create(
 }
 
 grpc_security_status grpc_ssl_server_security_connector_create(
-    grpc_exec_ctx *exec_ctx, const grpc_ssl_server_config *config,
+    grpc_exec_ctx *exec_ctx, const grpc_server_credentials *gsc,
     grpc_server_security_connector **sc) {
-  size_t num_alpn_protocols = grpc_chttp2_num_alpn_versions();
+  size_t num_alpn_protocols = 0;
   const char **alpn_protocol_strings =
-      gpr_malloc(sizeof(const char *) * num_alpn_protocols);
+      fill_alpn_protocol_strings(&num_alpn_protocols);
   tsi_result result = TSI_OK;
+  grpc_ssl_server_credentials *server_credentials =
+      (grpc_ssl_server_credentials *)gsc;
   grpc_ssl_server_security_connector *c;
-  size_t i;
-
-  for (i = 0; i < num_alpn_protocols; i++) {
-    alpn_protocol_strings[i] = grpc_chttp2_get_alpn_version_index(i);
-  }
 
-  if (config == NULL || config->num_key_cert_pairs == 0) {
-    gpr_log(GPR_ERROR, "An SSL server needs a key and a cert.");
+  if (server_credentials == NULL) {
+    gpr_log(GPR_ERROR, "Invalid SSL server credentials.");
     goto error;
   }
-  c = gpr_zalloc(sizeof(grpc_ssl_server_security_connector));
 
+  c = gpr_zalloc(sizeof(grpc_ssl_server_security_connector));
   gpr_ref_init(&c->base.base.refcount, 1);
   c->base.base.url_scheme = GRPC_SSL_URL_SCHEME;
   c->base.base.vtable = &ssl_server_vtable;
-  result = tsi_create_ssl_server_handshaker_factory_ex(
-      config->pem_key_cert_pairs, config->num_key_cert_pairs,
-      config->pem_root_certs, get_tsi_client_certificate_request_type(
-                                  config->client_certificate_request),
-      ssl_cipher_suites(), alpn_protocol_strings, (uint16_t)num_alpn_protocols,
-      &c->server_handshaker_factory);
-  if (result != TSI_OK) {
-    gpr_log(GPR_ERROR, "Handshaker factory creation failed with %s.",
-            tsi_result_to_string(result));
-    ssl_server_destroy(exec_ctx, &c->base.base);
-    *sc = NULL;
-    goto error;
-  }
   c->base.add_handshakers = ssl_server_add_handshakers;
+
+  c->certificate_store_fetcher = server_credentials->certificate_store_fetcher;
+  c->client_certificate_request =
+      server_credentials->config.client_certificate_request;
+
+  if (server_connector_has_cert_store_fetcher(c)) {
+    // Load initial credentials from certificate_store_fetcher:
+    if (!try_reload_ssl_server_credentials(c)) {
```

Comment:
a nit: function name is `reload` whereas we are actually loading the first time. i don't have good suggestions though